src/jemalloc.c

```diff
@@ -2363,13 +2363,25 @@ je_malloc(size_t size) {
 	assert(ind < SC_NBINS);
 	assert(size <= SC_SMALL_MAXCLASS);
 
-	uint64_t thread_allocated_after = thread_allocated_get(tsd) + usize;
-	assert(thread_allocated_next_event_fast_get(tsd) <=
-	    THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX);
-	if (unlikely(thread_allocated_after >=
-	    thread_allocated_next_event_fast_get(tsd))) {
+	/*
+	 * Use the unsafe getters since tsd may not be nominal, in which case
+	 * the fast_threshold will be set to 0.  This checks for events and tsd
+	 * non-nominal in a single branch.
+	 */
+	uint64_t *allocatedp = tsd_thread_allocatedp_get_unsafe(tsd);
+	/*
+	 * Not using relaxed atomic_load, because compilers right now can't tell
+	 * that this can be done as a memory comparison, and forces a load into
+	 * a register that hurts fast-path performance.
+	 */
+	uint64_t *fast_thresholdp =
+	    (uint64_t *)tsd_thread_allocated_next_event_fastp_get_unsafe(tsd);
```

Comment:
So by avoiding using the atomic load, the assembly code can be identical to before? What's the risk of not using the atomic load?

Reply:
Right. The asm output around this doesn’t change. If we do use atomic load, the fast threshold will be loaded into a register, instead of a comparison with memory directly.

With the current implementation I don’t think there’s any real risk. We are doing the same trick with tsd_state as well.

Reply:
Got it. For other places where we have less concern on the performance, should we guard the atomic loads and stores in an `#if JEMALLOC_ATOMIC_U64` and always provide an `#else` part?

Reply:
With @davidtgoldblatt 's help, the conclusion reached is, with a couple of memory fences the pattern here can be made safe w/o the `threshold_fast` being atomic. Stay tuned for David's comments (will be included in the diffs somewhere).